feature/game-state-preparation

letammyh opened the pull request:
ImageCache: added failedToCache to protocol
GameState: changed images property to [UserRecord:UIImage], added .preparing, .readyToPlay, .errorLoadingImages to GameState.Status
GameCoordinator: implemented
GameViewModel: implemented
GameWorkflow & VC: in progress (will be task of next feature branch)
MenuCoordinator: added GamePrepWorkflow
MenuViewModel: implemented to add activityIndicator functionality to menu screen while images are loading
Utilities: added Array and Dictionary extensions to shuffle array and search for key by value in dictionary

Commit:
ImageCache: added failedToCache to protocol
GameState: changed images property to [UserRecord:UIImage], added .preparing, .readyToPlay, .errorLoadingImages to GameState.Status
GameCoordinator: implemented
GameViewModel: implemented
GameWorkflow & VC: in progress (will be task of next feature branch)
MenuCoordinator: added GamePrepWorkflow
MenuViewModel: implemented to add activityIndicator functionality to menu screen while images are loading
Utilities: added Array and Dictionary extensions to shuffle array and search for key by value in dictionary

Nitewriter commented on NameGame/NameGame/Services/ImageCache.swift:
```diff
@@ -11,6 +11,7 @@ import UIKit
 
 protocol ImageCacheEventObserver: class {
     func didCache(image: UIImage, for userRecord: UserRecord)
+    func failedToCache(_ userRecord: UserRecord)
```
Change to: `failedToCacheImage(for userRecord: UserRecord)`

Nitewriter commented on NameGame/NameGame/Services/ImageCache.swift:
```diff
@@ -52,6 +53,7 @@ class ImageCache {
             }
         } catch {
             print("\(#function): \(error)")
```
Remove print statement.

Nitewriter commented on NameGame/NameGame/UI/Game/GameCoordinator.swift:
```diff
@@ -0,0 +1,79 @@
+        let controller = GameViewController.make()
+        let workflow = GameWorkflow(store: store)
+        
+        guard let gameState = store.state.gameState else {
```
Move this to your guard at the top of the `start` function. We shouldn't start if we don't have a game state.

Nitewriter commented on NameGame/NameGame/UI/Game/GameCoordinator.swift:
```diff
@@ -0,0 +1,79 @@
+    fileprivate let store: Store<AppState>
+    fileprivate let container: UINavigationController
+    private(set) var controller: GameViewController?
+    fileprivate var imageCache: ImageCache!
```
The game coordinator should need an imageCache. All images are provided in the game state.

Nitewriter commented on NameGame/NameGame/UI/Game/GameCoordinator.swift:
```diff
@@ -0,0 +1,79 @@
+        
+    }
+    
+    func inject(_ imageCache: ImageCache) {
```
Remove image cache.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,124 @@
+        guard let selectedImage = sender.backgroundImage(for: .normal) else {
+            return
+        }
+        guard let selectedUser = imageDictionary.key(for: selectedImage) else {
```
Combine this guard with the one above.

Nitewriter replied:
We should talk about another way to get the context of these buttons mapped back to our view model.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,124 @@
+        }
+        eventHandler.didGetUserChoice(selectedUser)
+        
+        // Wait a few seconds before displaying new question
```
We should talk about this one.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,124 @@
+    @IBOutlet var userButtons: [UIButton]!
+    @IBOutlet weak var scoreLabel: UILabel!
+    
+    fileprivate var imageDictionary: [UserRecord: UIImage]!
```
We should be able to access this image dictionary from the view model.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,124 @@
+        self.eventHandler = eventHandler
+    }
+    
+    func inject(_ imageDictionary: [UserRecord:UIImage]) {
```
I think we could get rid of the need to inject the image directory.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,77 @@
+struct GameViewModel {
+    
+    let choices: [UserRecord]?
+    let correctChoice: UserRecord?
```
Add the images to this model. Also, we should make these properties non-optional.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,77 @@
+    let questionAnsweredCorrectly: Bool?
+    let score: Int
+    
+    init() {
```
Remove this initializer.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,77 @@
+        score = 0
+    }
+    
+    init(_ state: AppState) {
```
Make this initializer optional `init?(_ state: AppState)`.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,77 @@
+    }
+    
+    init(_ state: AppState) {
+        guard let gameState = state.gameState else {
```
Guard for all the requirements we have, otherwise return nil.

Nitewriter commented on NameGame/NameGame/UI/Game/GameWorkflow.swift:
```diff
@@ -0,0 +1,73 @@
+        guard let gameState = store.state.gameState else {
+            return nil
+        }
+        let userRecords = gameState.userRecords.shuffled()
```
Talked through this one in RL. 😊

Nitewriter commented on NameGame/NameGame/UI/Game/GameWorkflow.swift:
```diff
@@ -0,0 +1,73 @@
+        }
+        
+        if playerChoice == store.state.gameState?.question.correctChoice {
+            store.dispatch(GameAction.setScore(gameState.score + 1))
```
Let's combine these two discreet action calls to one `AppAction.setGameState` call.

```swift
var newState = gameState
// Change all the things...
store.dispatch(AppAction.setGameState(newState)
```

Nitewriter commented on NameGame/NameGame/UI/Game/GameWorkflow.swift:
```diff
@@ -0,0 +1,73 @@
+    }
+    
+    func didGetNewQuestion() {
+        guard let newQuestion = generateNewQuestion() else {
```
Change this to an `if/else`.

```swift
if let newQuestion = generateNewQuestion() {
   store.dispatch(GameAction.setQuestion(newQuestion))
} else {
   store.dispatch(AppAction.endGame)
}
```

Nitewriter commented on NameGame/NameGame/UI/Game/GameWorkflow.swift:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+    
+    func didGetNewQuestion() {
```
I wonder if the name of this function is correct. It seems we are getting a new question or ending the game.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuCoordinator.swift:
```diff
@@ -52,13 +64,27 @@ extension MenuCoordinator: MenuWorkflowPresentationEventObserver {
     func presentGameCoordinator() {
-        // TODO
+        let coordinator = GameCoordinator(store: store, container: container)
+        coordinator.inject(imageCache)
```
Drop the image cache.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+    }
+    
+    init(_ state: AppState) {
+        guard let gameState = state.gameState else {
```
We should talk about how to get rid of this guard.

Nitewriter commented on NameGame/NameGame/UI/Menu/Wokflows/GamePrepWorkflow.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    private func loadGameStateImages(with userRecords: [UserRecord]) {
+        guard
```
Style change:

```swift
guard var images = store.state.gameState?.images else {
    return
}
```

Nitewriter replied:
I think we may also want a `preconditionFailure()` here so we will know quickly we are calling this when we shouldn't.

Nitewriter commented on NameGame/NameGame/UI/Menu/Wokflows/GamePrepWorkflow.swift:
```diff
@@ -0,0 +1,89 @@
+        guard var images = store.state.gameState?.images else {
+            return
+        }
+        images[userRecord] = image
```
Style change: add a new line after the guard.

Nitewriter commented on NameGame/NameGame/UI/UserGrid/UserGridCoordinator.swift:
```diff
@@ -17,7 +17,7 @@ final class UserGridCoordinator: Coordinator {
     private var imageCache: ImageCache!
     
-    weak var controllerEventObserver: CoordinatorEventObserver?
+    weak var coordinatorEventObserver: CoordinatorEventObserver?
```
Maybe shorten this to `eventObserver`.

Nitewriter commented on NameGame/NameGame/Utilities/Collections/ArrayExtensions.swift:
```diff
@@ -0,0 +1,25 @@
+extension Array {
+    
+    private mutating func shuffle() {
+        for _ in 0..<self.count {
```
Talked about this in RL.

Nitewriter commented on NameGame/NameGame/Utilities/Collections/DictionaryExtensions.swift:
```diff
@@ -0,0 +1,22 @@
+
+import Foundation
+
+extension Dictionary where Value: Equatable {
```
I think we may be able to drop this extension.

Nitewriter reviewed (changes requested):
As you work through the feed back, feel free to grab me for discussion.

Commit:
Headshot: didn't initialize user record if their headshot image was default WT image
UserRecordService: removed shared instance
GameState: added finishedPlaying status, added didFinishPlaying property
EndGame: implemented VC, ViewModel, Workflow, and Coordinator for EndGame screen
GameViewModel: made it an optional
MenuViewController: added code for animation

Nitewriter commented on NameGame/NameGame/Services/Model/Headshot.swift:
```diff
@@ -58,6 +58,11 @@ struct Headshot {
             throw ParsingError.missingKey(CodingKey.width.rawValue)
         }
         
+        // Don't initialize a user record if their image is the default WT image
```
Nice!

Nitewriter replied:
Maybe change to a guard:

```swift
guard urlString != "//images.contentful.com/3cttzl4i3k1h/5ZUiD3uOByWWuaSQsayAQ6/c630e7f851d5adb1876c118dc4811aed/featured-image-TEST1.png" else {
    throw ParsingError.hasDefaultImage
}
```

Nitewriter commented on NameGame/NameGame/UI/EndGame/EndGameViewController.swift:
```diff
@@ -0,0 +1,45 @@
+class EndGameViewController: UIViewController {
+
+    @IBOutlet weak var scoreLabel: UILabel!
+    private var finalScore: Int!
```
Default this to `0` instead of force unwrapping.

```swift
private var finalScore = 0
```

Nitewriter commented on NameGame/NameGame/UI/EndGame/EndGameViewController.swift:
```diff
@@ -0,0 +1,45 @@
+        self.eventHandler = eventHandler
+    }
+    
+    func inject(_ finalScore: Int) {
```
Can we subscribe to the `AppStore` to get the final score data?

Nitewriter commented on NameGame/NameGame/UI/Game/GameCoordinator.swift:
```diff
@@ -0,0 +1,105 @@
+        }
+        
+        store.subscribe(controller) { state in
+            state.select(GameViewModel.init) }
```
Style; move that closing curly brace to the next line.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    @IBOutlet var userButtons: [UIButton]!
+    @IBOutlet weak var scoreLabel: UILabel!
+    
+    fileprivate var imageDictionary: [UserRecord: UIImage]!
```
We should be able to remove the `imageDictionary` property from the `GameViewController`.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    fileprivate var currentCorrectChoiceIndex = 0
+    fileprivate var currentSelectedChoiceIndex = 0
+    
+    fileprivate let green = UIColor(red: 74.0/255.0, green: 226.0/255.0, blue: 59.0/255.0, alpha: 1)
```
Maybe lift these to a color palette struct.

```swift
struct ColorPalette {
    static let green: UIColor {
         return UIColor(red: 74.0/255.0, green: 226.0/255.0, blue: 59.0/255.0, alpha: 1)
    }

    ...
}
```

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    weak var lifecycleObserver: ViewLifecycleObserver? = nil
+    fileprivate var eventHandler: GameViewEventHandler!
+    
+    fileprivate var currentCorrectChoiceIndex = 0
```
This index data should live in the `GameViewModel`.

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,164 @@
+            }
+        }
+        
+        nameLabel.text = viewModel.correctChoice.firstName + " " + viewModel.correctChoice.lastName
```
Extra credit; check out this tool instead of string concatenation. https://developer.apple.com/documentation/foundation/personnamecomponentsformatter

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewController.swift:
```diff
@@ -0,0 +1,164 @@
+            userButtons[currentSelectedChoiceIndex].layer.borderWidth = 3.0
+        }
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1), execute: {
```
Style; use trailing closure syntax here.

```swift
DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
    ...
}
```

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,66 @@
+    let imageDictionary: [UserRecord: UIImage]
+    let questionAnsweredCorrectly: Bool?
+    let score: Int
+//    let remainingQuestions: Int
```
Do we need to keep this commented property?

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        let decreaseSize = CABasicAnimation(keyPath: "bounds")
+        decreaseSize.delegate = self
+        decreaseSize.duration = 1.0
+        decreaseSize.fromValue = NSValue(cgRect: mask!.bounds)
```
Can we guard or pass the mask to this function instead of force unwrapping?

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    private func animateIncreaseSize() {
+        animation = CABasicAnimation(keyPath: "bounds")
+        animation.duration = 2
+        animation.fromValue = NSValue(cgRect: mask!.bounds)
```
Same thing here.

Nitewriter commented on NameGame/NameGame/UI/Menu/Wokflows/GamePrepWorkflow.swift:
```diff
@@ -0,0 +1,111 @@
+        }
+        
+        let mattUserRecords = userRecords.filter { $0.firstName.contains("Mat") }
+        GamePrepWorkflow.answerQueueSize = mattUserRecords.count
```
Should we handle the possibility of there being no `Mat(t)`s in the data set?

Nitewriter reviewed (changes requested):
Wow! This is really coming together Tammy. Just a few comments for you.

Commit:
Headshot: added guard statement
EndGameVC: subscribed to store to render final score
GameViewModel: moved correctChoiceIndex here, used PersonNameComponents
MenuVC: made mask a parameter to remove the forced unwrapping
GamePrepWorkflow: handled case if there are no Mat(t) user records
ColorPalette: implemented custom colors here

Nitewriter commented on NameGame/NameGame/UI/Game/GameViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+        correctChoice = gameState.question.correctChoice
+        correctChoiceIndex = choices.index(of: correctChoice) ?? 0
+        
+        correctChoiceNameComponents = PersonNameComponents()
```
Very nice! 😉

Nitewriter reviewed (approved):
LGTM! 💰
